### DISCUSSION TITLE
Content-Security-Policy „default-src 'self';“ and the admin console

### DISCUSSION THREAD

**Discussion by esunke** [keycloak/keycloak#11878] [Archive]
Hi,
 I am using Keycloak 16.1.1 (Bitname docker image). A spart of security checks, we are forced to enable several content security policies. When I enable them, I cannot  access the admin console UI  any more. I found some discussions that stated that I should use the Keycloak v2 theme So I enabled the admin2 feature and activated the new theme, However, the admin console is still not usable. I get several issues related to the CSP. Is there any way to enable the policies and also use the admin UI?
Thanks, Eik

**Comment by darius-m** [comment]
The `default-src: 'self'` policy should instruct the browser to automatically use this policy for any type of resource (e.g., images, javascript, stylesheets) that does not have an explicit policy. `'self'` means that all data should be stored on the keycloak server itself.

The admin console probably requires outside dependencies at least for CSS files (it relies on 3rd party frameworks). Without seeing the issues, I will assume that there are some errors with downloading the CSS stylesheets or javascript scripts. From what I understand (I am also interested in input from someone more knowledgeable), you have two options:
  - host all frameworks on the keycloak servers (this will probably increase the load on the servers unless you also use a caching reverse proxy);
  - host a CDN server that you trust with the frameworks and change the references used by the themes to your CDNs.

This being said, I do not know how to change how themes are built, or if having no external dependencies is an currently an option.

**Reply by esunke** [reply]
Thanks for the reply.
I just checked in more detail.
The issues coming up are mainly based on inline scripts and inline styles. 

"Refused to apply inline style" and "Refused to execute inline script".

To be sure I just tested with the 18.0.0 version of keycloak. But still, the directive cannot be used. 
![image](https://user-images.githubusercontent.com/29062113/167247094-fd4abd70-890c-4901-a59e-9c80c6e1bb03.png)

Since we are talking about an security relevant part of an architecture, this is something that should be addressed and fixed.

**Reply by darius-m** [reply]
Inline scripts are implicitly blocked by the `self` policy since it explicitly requires the scripts to be downloaded from the server as separate files. Fixing this in the code would be ideal (I would also prefer it), but impossible depending on how the framework used by the interface generates the pages.

A possible workaround would be adding the `unsafe-inline` or a nonce value to the [script-src](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src#unsafe_inline_script), but it may not be acceptable for everyone.

**Reply by darius-m** [reply]
There also seems to be an issue with styling, so similar options should be set for [style-src](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src#unsafe_inline_styles). In this case, I believe separating styling and moving it to other files is harder, since frameworks may prefer adding styles inline with the HTML element.  Something like `<div style="display: none"` to hide elements, for example.
